p2p/pex/reactor.go

```diff
@@ -0,0 +1,189 @@
+package pex
+
+import (
+	"fmt"
+
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/libs/service"
+	"github.com/tendermint/tendermint/p2p"
+	protop2p "github.com/tendermint/tendermint/proto/tendermint/p2p"
+)
+
+var (
+	_ service.Service = (*ReactorV2)(nil)
+	_ p2p.Wrapper     = (*protop2p.Message)(nil)
+)
+
+const (
+	maxAddresses uint16 = 100
+)
+
+// ReactorV2 is a PEX reactor for the new P2P stack. The legacy reactor
+// is Reactor.
+//
+// FIXME: Rename this when Reactor is removed, and consider moving to p2p/.
+type ReactorV2 struct {
+	service.BaseService
+
+	peerManager *p2p.PeerManager
+	pexCh       *p2p.Channel
+	peerUpdates *p2p.PeerUpdatesCh
+	closeCh     chan struct{}
+}
+
+// NewReactor returns a reference to a new reactor.
+func NewReactorV2(
+	logger log.Logger,
+	peerManager *p2p.PeerManager,
+	pexCh *p2p.Channel,
+	peerUpdates *p2p.PeerUpdatesCh,
+) *ReactorV2 {
+	r := &ReactorV2{
+		peerManager: peerManager,
+		pexCh:       pexCh,
+		peerUpdates: peerUpdates,
+		closeCh:     make(chan struct{}),
+	}
+
+	r.BaseService = *service.NewBaseService(logger, "PEX", r)
+	return r
+}
+
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed.
+func (r *ReactorV2) OnStart() error {
+	go r.processPexCh()
+	go r.processPeerUpdates()
+	return nil
+}
+
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *ReactorV2) OnStop() {
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
+
+	// Wait for all p2p Channels to be closed before returning. This ensures we
+	// can easily reason about synchronization of all p2p Channels and ensure no
+	// panics will occur.
+	<-r.pexCh.Done()
+	<-r.peerUpdates.Done()
+}
+
+// handlePexMessage handles envelopes sent from peers on the PexChannel.
+func (r *ReactorV2) handlePexMessage(envelope p2p.Envelope) error {
+	logger := r.Logger.With("peer", envelope.From)
+
+	// FIXME: We may want to add DoS protection here, by rate limiting peers and
+	// only processing addresses we actually requested.
+	switch msg := envelope.Message.(type) {
+	case *protop2p.PexRequest:
+		endpoints := r.peerManager.Advertise(envelope.From, maxAddresses)
+		resp := &protop2p.PexAddrs{Addrs: make([]protop2p.NetAddress, 0, len(endpoints))}
+		for _, endpoint := range endpoints {
+			// FIXME: This shouldn't rely on NetAddress.
```

Comment:
Why should this not rely on `NetAddress`?

Reply:
Because `NetAddress` is an artifact of the legacy P2P stack. The new stack uses `PeerAddress` and `Endpoint`.